README.md
# oScope, the OSC scope

This is a command line utility for sending and receiving [Open Sound Control](https://en.wikipedia.org/wiki/Open_Sound_Control) messages. This can be useful for debugging and testing.

It's currently in an early release—if you see something missing, [please reach out](https://github.com/mindofmatthew/oscope/issues)!

## Installation

You can use it two ways. Both require you to have Node.js installed.

If you want to install the tool so you can have it available in your path, run:

```bash
npm install -g oscope

#Then run it with
oscope <command> [options...]
```

A simpler solution is to run it with NPX like so:

```bash
npx oscope <command> [options...]
```

## Usage

Currently, oscope can be used with two commands, `listen` and `talk`. Both take a UDP address as an argument.

### Address Patterns

Addresses consist of a hostname and a port such as `127.0.0.1:8000` or `localhost:1234`. A port by itself (such as `8888`) is also allowed.

### `listen`

```bash
oscope listen <address>
```

This opens a connection at the given local address and prints the contents of any OSC messages received. If only the port is specified, then it listens for messages sent to that port on all addresses.

### `talk`

```bash
oscope talk <address>
```

This opens a connection to a piece of software at a given remote address. You can use the command line to send messages to that program and then view any messages sent in response on the console.


HISTORY.md
# History

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2021-05-11

### Added

- Initial implementation of functions for sending and receiving OSC messages.


src/cli.ts
#!/usr/bin/env node

import { createSocket, Socket } from "dgram";
import { createInterface } from "readline";

import { bold, dim, blue, red, inverse } from "chalk";

import { parse, message as oscMessage } from "./osc";
import {
  OSCArgumentValueList,
  OSCArgumentTagList,
  OSCBundle,
  OSCMessage,
} from "./types";

let [action = "", ...args] = process.argv.slice(2);

switch (action.toLowerCase()) {
  case "help":
    break;
  case "talk":
    talk(args);
    break;
  case "listen":
    listen(args);
    break;
  case "snoop":
    snoop(args);
    break;
  default:
    //error
    break;
}

// Sub-programs
function help() {}

function talk(args: string[]) {
  let [address, port, addressType] = parseAddress(args[0]);

  let socket = createSocket(addressType);

  socket.connect(port, address);

  socket.on("connect", () => {
    let { address, port } = socket.remoteAddress();
    console.log(inverse(center(`Sending OSC to ${address}:${port}`)));

    input.prompt();
  });

  socket.on("message", (message, { address, port }) => {
    printDatagram(message, address, port);
  });

  let input = createInterface({ input: process.stdin, output: process.stdout });

  input.on("line", (line) => {
    try {
      // Parse message
      let [oscAddress, oscArgs = ""] = line.trim().split(/\s+(.*)/);

      let oscArgValues = [];

      while (oscArgs.length > 0) {
        let arg: string;
        let match: RegExpMatchArray | null;

        if ((match = oscArgs.match(/^([+-]?\d+)(?:$|\s+(.*)$)/))) {
          [, arg, oscArgs = ""] = match;
          oscArgValues.push({ i: parseInt(arg) });
        } else if (
          (match = oscArgs.match(
            /^([+-]?(?:\d+\.\d*|\.\d+|\d+f))(?:$|\s+(.*)$)/
          ))
        ) {
          [, arg, oscArgs = ""] = match;
          oscArgValues.push({ f: parseFloat(arg) });
        } else if ((match = oscArgs.match(/^"([^"]*)"(?:$|\s+(.*)$)/))) {
          [, arg, oscArgs = ""] = match;
          oscArgValues.push({ s: arg });
        } else {
          throw Error(`Didn't recognize character "${oscArgs[0]}"`);
        }
      }

      socket.send(oscMessage(oscAddress, ...oscArgValues));
    } catch (e) {
      console.log(red(e.message));
    }
  });
}

function listen(args: string[]) {
  let [address, port, addressType] = parseAddress(args[0]);

  let socket = createSocket(addressType);

  socket.on("listening", () => {
    const info = socket.address();

    console.log(
      inverse(center(`Listening for OSC on ${info.address}:${info.port}`))
    );
  });

  socket.on("message", (message, { address, port }) => {
    printDatagram(message, address, port);
  });

  socket.bind(port, address);
}

function snoop(args: string[]) {
  let [address1, port1] = parseAddress(args[0]);
  let [address2, port2] = parseAddress(args[1]);

  console.log(address2, port2);

  let socket1 = createSocket("udp4");

  let remotes = new Map<number, Socket>();

  socket1.on("listening", () => {
    console.log(
      inverse(`  Snooping between ${socket1.address().port} and ${port2}  `)
    );
  });

  socket1.on("message", (message, rinfo) => {
    let remote: Socket;

    if (remotes.has(rinfo.port)) {
      remote = remotes.get(rinfo.port) as Socket;
      remote.send(message);
    } else {
      remote = createSocket("udp4");
      remote.connect(port2, address2);

      remote.on("connect", () => {
        remote.send(message);
        remotes.set(rinfo.port, remote);
      });

      remote.on("message", (message, rinfo2) => {
        socket1.send(message, rinfo.port, rinfo.address);
      });
    }

    printDatagram(message, rinfo.address, rinfo.port);
  });

  socket1.bind(port1, address1);
}

function printDatagram(packet: Uint8Array, address: string, port: number) {
  try {
    console.log(
      "\n" + blue(`${address}:${port} (received ${printTime(new Date())})`)
    );

    let result = parse(packet);
    printPacket(result);
  } catch (error) {
    console.log(red(error.message));
  }
}

const NTPOffset = 2208988800;

function printPacket(packet: OSCBundle | OSCMessage, indent = 0) {
  if ("packets" in packet) {
    let {
      ntpTime: [seconds, fracSeconds],
      packets,
    } = packet;
    let time = new Date((seconds - NTPOffset + fracSeconds / 2 ** 32) * 1000);
    console.log("  ".repeat(indent) + dim(`Bundle (${printTime(time)})`));

    for (let subPacket of packets) {
      printPacket(subPacket, indent + 1);
    }
  } else {
    let { address, args, argTypes } = packet;
    console.log(
      "  ".repeat(indent) + `${bold(address)} ${serializeArgs(args, argTypes)}`
    );
  }
}

// Utils
function parseAddress(input: string): [string | undefined, number, 'udp4' | 'udp6'] {
  let match: RegExpMatchArray | null;
  let address: string, portString: string;
  let addressType: 'udp4' | 'udp6';
  if (match = input.match(/^(?:([0-9a-z\.]*):)?(\d+)$/)) {
    [, address, portString] = match;
    addressType = 'udp4';
  } else if (match = input.match(/^(?:\[([0-9a-f:]*)\]:)?(\d+)$/)) {
    [, address, portString] = match;
    addressType = 'udp6';
  } else {
    throw Error(`Unrecognized address: "${input}"`);
  }

  return [address || undefined, parseInt(portString), addressType];
}

function serializeArgs(args: OSCArgumentValueList, types: OSCArgumentTagList) {
  return args
    .map((arg, i) => {
      if (typeof arg === "number") {
        if (types[i] === "f") {
          return arg.toFixed(3);
        } else {
          return arg.toString();
        }
      } else if (typeof arg === "string") {
        return `"${arg}"`;
      } else if (arg instanceof Uint8Array) {
        return `<Blob (${arg.length}B)>`;
      }
    })
    .join(" ");
}

const dateFormatter = new Intl.DateTimeFormat(undefined, {
  year: "numeric",
  month: "numeric",
  day: "numeric",
});

const timeFormatter = new Intl.DateTimeFormat("en-US", {
  hour: "numeric",
  minute: "numeric",
  second: "numeric",
  hour12: false,
});

function printTime(time: Date) {
  return (
    dateFormatter.format(time) +
    " " +
    timeFormatter.format(time) +
    "." +
    time.getMilliseconds().toString().padStart(3, "0")
  );
}

function center(text: String) {
  // Default padding
  let pad: number = 2;

  if (process.stdout.isTTY) {
    let [width] = process.stdout.getWindowSize();
    pad = Math.max(pad, (width - text.length) / 2);
  }

  return " ".repeat(Math.ceil(pad)) + text + " ".repeat(Math.floor(pad));
}


src/cli/index.ts
#!/usr/bin/env node

import { createSocket, Socket } from "dgram";
import { createInterface } from "readline";

import { bold, dim, blue, red, inverse } from "chalk";

import { parseAddress, printAddress } from "./address";
import { errAddressInUse } from "./errors";

import { parse, message as oscMessage } from "../osc/osc";
import {
  OSCArgumentValueList,
  OSCArgumentTagList,
  OSCBundle,
  OSCMessage,
} from "../osc/types";

let [action = "", ...args] = process.argv.slice(2);

switch (action.toLowerCase()) {
  case "talk":
    talk(args);
    break;
  case "listen":
    listen(args);
    break;
  case "help":
    help();
    break;
  default:
    // Check for a different action
    if (action != "") {
      console.log(`I don't understand the command "${action}"\n`);
    }

    console.log("Supported commands:\n");

    help(); // Print help info

    break;
}

// Sub-programs
function help() {
  console.log(bold("oscope listen <address>"));
  console.log("  Open a UDP port on <address> and print received messages.\n");

  console.log(bold("oscope talk <address>"));
  console.log(
    "  Open a text prompt for sending messages to another piece of software listening on <address>.\n"
  );

  console.log(bold("oscope help"));
  console.log("  Print this help information.\n");

  console.log(`(oscope version ${process.env.npm_package_version})`);
}

function talk(args: string[]) {
  let { address, port, family } = parseAddress(args[0]);

  let socket = createSocket(family === "IPv4" ? "udp4" : "udp6");

  socket.connect(port, address || undefined);

  socket.on("connect", () => {
    let info = socket.remoteAddress();
    console.log(inverse(center(`Sending OSC to ${printAddress(info)}`)));

    input.prompt();
  });

  socket.on("message", (message, { address, port }) => {
    printDatagram(message, address, port);
  });

  let input = createInterface({ input: process.stdin, output: process.stdout });

  input.on("line", (line) => {
    try {
      // Parse message
      let [oscAddress, oscArgs = ""] = line.trim().split(/\s+(.*)/);

      let oscArgValues = [];

      while (oscArgs.length > 0) {
        let arg: string;
        let match: RegExpMatchArray | null;

        if ((match = oscArgs.match(/^([+-]?\d+)(?:$|\s+(.*)$)/))) {
          [, arg, oscArgs = ""] = match;
          oscArgValues.push({ i: parseInt(arg) });
        } else if (
          (match = oscArgs.match(
            /^([+-]?(?:\d+\.\d*|\.\d+|\d+f))(?:$|\s+(.*)$)/
          ))
        ) {
          [, arg, oscArgs = ""] = match;
          oscArgValues.push({ f: parseFloat(arg) });
        } else if ((match = oscArgs.match(/^"([^"]*)"(?:$|\s+(.*)$)/))) {
          [, arg, oscArgs = ""] = match;
          oscArgValues.push({ s: arg });
        } else {
          throw Error(`Didn't recognize character "${oscArgs[0]}"`);
        }
      }

      socket.send(oscMessage(oscAddress, ...oscArgValues));
    } catch (e) {
      console.log(red(e.message));
    }
  });
}

function listen(args: string[]) {
  let { address, port, family } = parseAddress(args[0]);

  let socket = createSocket(family === "IPv4" ? "udp4" : "udp6");

  socket.on("listening", () => {
    const info = socket.address();

    console.log(
      inverse(center(`Listening for OSC on ${info.address}:${info.port}`))
    );
  });

  socket.on("message", (message, { address, port }) => {
    printDatagram(message, address, port);
  });

  socket.on("error", (err) => {
    if (errAddressInUse(err)) {
      console.log(
        red(
          `Error: Another program is already listening to the UPD socket ${err.address}:${err.port}`
        )
      );
    } else {
      // Unknown error
      console.log(err);
    }
  });

  socket.bind(port, address);
}

function snoop(args: string[]) {
  let { address: address1, port: port1 } = parseAddress(args[0]);
  let { address: address2, port: port2 } = parseAddress(args[1]);

  console.log(address2, port2);

  let socket1 = createSocket("udp4");

  let remotes = new Map<number, Socket>();

  socket1.on("listening", () => {
    console.log(
      inverse(`  Snooping between ${socket1.address().port} and ${port2}  `)
    );
  });

  socket1.on("message", (message, rinfo) => {
    let remote: Socket;

    if (remotes.has(rinfo.port)) {
      remote = remotes.get(rinfo.port) as Socket;
      remote.send(message);
    } else {
      remote = createSocket("udp4");
      remote.connect(port2, address2);

      remote.on("connect", () => {
        remote.send(message);
        remotes.set(rinfo.port, remote);
      });

      remote.on("message", (message, rinfo2) => {
        socket1.send(message, rinfo.port, rinfo.address);
      });
    }

    printDatagram(message, rinfo.address, rinfo.port);
  });

  socket1.bind(port1, address1);
}

function printDatagram(packet: Uint8Array, address: string, port: number) {
  try {
    console.log(
      "\n" + blue(`${address}:${port} (received ${printTime(new Date())})`)
    );

    let result = parse(packet);
    printPacket(result);
  } catch (error) {
    console.log(red(error.message));
  }
}

const NTPOffset = 2208988800;

function printPacket(packet: OSCBundle | OSCMessage, indent = 0) {
  if ("packets" in packet) {
    let {
      ntpTime: [seconds, fracSeconds],
      packets,
    } = packet;
    let time = new Date((seconds - NTPOffset + fracSeconds / 2 ** 32) * 1000);
    console.log("  ".repeat(indent) + dim(`Bundle (${printTime(time)})`));

    for (let subPacket of packets) {
      printPacket(subPacket, indent + 1);
    }
  } else {
    let { address, args, argTypes } = packet;
    console.log(
      "  ".repeat(indent) + `${bold(address)} ${serializeArgs(args, argTypes)}`
    );
  }
}

function serializeArgs(args: OSCArgumentValueList, types: OSCArgumentTagList) {
  return args
    .map((arg, i) => {
      if (typeof arg === "number") {
        if (types[i] === "f") {
          return arg.toFixed(3);
        } else {
          return arg.toString();
        }
      } else if (typeof arg === "string") {
        return `"${arg}"`;
      } else if (arg instanceof Uint8Array) {
        return `<Blob (${arg.length}B)>`;
      }
    })
    .join(" ");
}

const dateFormatter = new Intl.DateTimeFormat(undefined, {
  year: "numeric",
  month: "numeric",
  day: "numeric",
});

const timeFormatter = new Intl.DateTimeFormat("en-US", {
  hour: "numeric",
  minute: "numeric",
  second: "numeric",
  hour12: false,
});

function printTime(time: Date) {
  return (
    dateFormatter.format(time) +
    " " +
    timeFormatter.format(time) +
    "." +
    time.getMilliseconds().toString().padStart(3, "0")
  );
}

function center(text: String) {
  // Default padding
  let pad: number = 2;

  if (process.stdout.isTTY) {
    let [width] = process.stdout.getWindowSize();
    pad = Math.max(pad, (width - text.length) / 2);
  }

  return " ".repeat(Math.ceil(pad)) + text + " ".repeat(Math.floor(pad));
}
